apps/src/lab2/responseValidators.ts

```diff
@@ -1,26 +1,45 @@
 import {ResponseValidator} from '@cdo/apps/util/HttpClient';
-import {BlocklySource, ProjectSources} from './types';
+import {BlocklySource, LevelProperties, ProjectSources} from './types';
 
 export const SourceResponseValidator: ResponseValidator<
   ProjectSources
 > = response => {
-  const projectSources = response as ProjectSources;
-  if (!projectSources.source) {
+  if (!response.source) {
     throw new ValidationError('Missing required field: source');
   }
 
   // Currently only Blockly JSON sources are supported.
   let blocklySource;
   try {
-    blocklySource = JSON.parse(projectSources.source) as BlocklySource;
+    blocklySource = JSON.parse(response.source as string) as BlocklySource;
   } catch (e) {
     throw new ValidationError('Error parsing JSON: ' + e);
   }
   if (blocklySource.blocks === undefined) {
     throw new ValidationError('Missing required field: blocks');
   }
 
-  return projectSources;
+  return response as unknown as ProjectSources;
+};
+
+export const LevelPropertiesValidator: ResponseValidator<
+  LevelProperties
+> = response => {
+  if (!response.appName) {
+    throw new ValidationError('Missing required field: appName');
+  }
+
+  // Convert stringified booleans to actual booleans.
+  for (const key of Object.keys(response)) {
+    if (response[key] === 'true') {
+      response[key] = true;
+    }
+    if (response[key] === 'false') {
+      response[key] = false;
+    }
+  }
+
+  return response as unknown as LevelProperties;
```

Comment:
Could you explain this syntax `return response as unknown as LevelProperties`?

Reply:
Yeah, this is a bit confusing 🙂 `response` is of type `Record<string, unknown>` and TypeScript doesn't believe that a `Record<string, unknown>` type can be assigned to `LevelProperties` because there's no overlap in the types. However, we do know that this is response is indeed a `LevelProperties` type because we know which API we're calling, and it's passed validation. So in order to type cast it, we need to first cast it to `unknown` which removes all typing, and then cast it to `LevelProperties`. More on the `unknown` type here: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-0.html#new-unknown-top-type. Basically it's a way to say that this type could be anything, but we have to first narrow the type (by casting it here) before treating it as a specific type.